src/vos/vos_dtx_iter.c

```diff
@@ -112,24 +135,42 @@ dtx_iter_next(struct vos_iterator *iter)
 	D_ASSERT(iter->it_type == VOS_ITER_DTX);
 
 	while (1) {
-		rc = dbtree_iter_next(oiter->oit_hdl);
-		if (rc != 0)
-			break;
-
-		d_iov_set(&rec_iov, NULL, 0);
-		rc = dbtree_iter_fetch(oiter->oit_hdl, NULL, &rec_iov, NULL);
-		if (rc != 0)
-			break;
-
-		D_ASSERT(rec_iov.iov_len == sizeof(struct vos_dtx_act_ent));
-		dae = (struct vos_dtx_act_ent *)rec_iov.iov_buf;
+		if (oiter->oit_linear) {
+			if (oiter->oit_cur == NULL)
+				D_GOTO(out, rc = -DER_NONEXIST);
+
+			if (oiter->oit_cur->dae_link.next ==
+			    &oiter->oit_cont->vc_dtx_act_list) {
+				oiter->oit_cur = NULL;
+				D_GOTO(out, rc = -DER_NONEXIST);
+			}
+
+			dae = oiter->oit_cur =
+				d_list_entry(oiter->oit_cur->dae_link.next,
+					     struct vos_dtx_act_ent, dae_link);
+		} else {
+			rc = dbtree_iter_next(oiter->oit_hdl);
+			if (rc != 0)
+				goto out;
+
+			d_iov_set(&rec_iov, NULL, 0);
+			rc = dbtree_iter_fetch(oiter->oit_hdl, NULL,
+					       &rec_iov, NULL);
+			if (rc != 0)
+				goto out;
+
+			D_ASSERT(rec_iov.iov_len ==
+				 sizeof(struct vos_dtx_act_ent));
+			dae = (struct vos_dtx_act_ent *)rec_iov.iov_buf;
```

Comment:
You probably don't need an assert here.

Reply:
The DTX iteration can be used for cleanup stale DTX entries. Here the check is mainly to verify that the "stale" entry is not caused by local corruption. So I prefer to keep such assertion.

Reply:
I just means that iov_buf will be a "void *" which alias to anything, I was thinking about the cast rather than the assert.

Looking at this now I typed the wrong thing, the assert makes sense as you say, but the cast could be removed.  Not worth re-pushing for alone but it would be easier to read and safer without it.